### PR TITLE
Extend user with photo authentication

### PR DESCRIPTION
**Pull request by eshaibu**
### What PR does
Extend Syncano user functionality to use aws photo rekognition for face authentication

#### face_register
- Expected a  user sends image (I expect on the client side the user will be asked to confirm password). 
- The endpoint takes parameters of username, password, image, collectionId
- Endpoint validates user exist.
- If user exist validate image contains just one face (Image with more than person rejected).
- Search image ensure no other user with same face.
- Index image and also set externalImageId on the indexed face.
- Update user schema extended field `face_auth` to true and `external_image_id `to indexed image externalImageId.
#### face_login
- Use image to login if face_registered to user account
- Checks externalImageId of search match if found in users
- Return user details
#### remove_face_auth
- Remove face authentication for particular user
- Validates user exist
- Checks face authentication enabled for user account
- Ensure face image parameter for logged in user (token supplied as parameter)
- If image for user delete faces associated to account and update schema to `face_auth` to false and `external_image_id` to empty
#### verify_face_auth
- Remove face authentication for particular user
- Verifies if face authentication enabled for user account

### CODE REVIEW

File: socket.yml [hunk 1]
```diff
@@ -102,33 +122,28 @@ endpoints:
           description: Success
           example: |
             {
-            "statusCode": 200,
-            "userId": "12345",
+              message: "User account registered for face authentication."
```
Comment (adamwardecki) [inline socket.yml]: "User face registered for face authentication."

File: src/face_login.js [hunk 2]
```diff
@@ -3,31 +3,45 @@ import Syncano from 'syncano-server';
 import Rekognition from './utils/Rekognition';
 
 export default (ctx) => {
-  const {response} = Syncano(ctx);
+  const {response, users } = Syncano(ctx);
 
   const awsRekognitionClass = new Rekognition(ctx.config);
 
   const {collectionId, image} = ctx.args;
   const bucketName = (ctx.args.bucketName || ctx.args.bucketName.trim() !== '')
```
Comment (adamwardecki) [inline src/face_login.js]: This assumes that the bucketName param is sent in the request. It should be optional.

File: src/face_register.js [hunk 3]
```diff
@@ -1,45 +1,110 @@
...
 
 export default (ctx) => {
-  const {response} = Syncano(ctx);
+  const { users, response } = Syncano(ctx);
+
+  const {
+    username, password, collectionId, image, bucketName
+  } = ctx.args;
+
+  const s3bucket = (bucketName || bucketName.trim() !== '') ? bucketName : null;
```
Comment (adamwardecki) [inline src/face_register.js]: This assumes that the bucketName param is sent in the request. It should be optional.

File: src/face_login.js [hunk 4]
```diff
@@ -3,31 +3,45 @@ import Syncano from 'syncano-server';
 import Rekognition from './utils/Rekognition';
 
 export default (ctx) => {
-  const {response} = Syncano(ctx);
+  const {response, users } = Syncano(ctx);
```
Comment (adamwardecki) [inline src/face_login.js:6]: const {response, users, bucketName } = Syncano(ctx);

File: src/face_login.js [hunk 5]
```diff
@@ -3,31 +3,45 @@ import Syncano from 'syncano-server';
...
+
+  /**
+   * Get user based on face match externalID
+   * @param {object} res
+   * @returns {Promise.<*>} promise
+   */
+  const getUser = (res) => {
+    return users.where('external_image_id', res.FaceMatches[0].Face.ExternalImageId)
+      .firstOrFail()
+      .then((data) => {
```
Comment (adamwardecki) [inline src/face_login.js]: you can unpack the res data, i.e. { user_key, username }

File: src/utils/Rekognition.js [hunk 6]
```diff
@@ -117,24 +151,39 @@ class Rekognition {
...
    *
    * @param {string} collectionId
    * @param {string} image
    * @param {?string} bucket
    * @param {number} threshold
-   * @returns {Promise} object response
+   * @returns {Promise}
    */
-  searchFacesByImage(collectionId, image, bucket = null, threshold = 90) {
+  async searchFacesByImage(collectionId, image, bucket = null, threshold = 95) {
```
Comment (adamwardecki) [inline src/utils/Rekognition.js]: As we discussed, threshold should be set with config

File: src/face_register.js [hunk 7]
```diff
@@ -1,45 +1,110 @@
...
+      });
+  };
+
+  axios.post(AUTH_URL, { username, password },
+    {
+      headers: { 'Content-Type': 'application/json', 'X-API-KEY': ctx.meta.token }
     })
+    .then(searchUserFace)
+    .then(indexUserFace)
+    .then(handleSuccessfulIndexImage)
```
Comment (adamwardecki) [inline src/face_register.js]: It handles both success and failure so maybe function should be called `handleIndexImage`

File: src/verify_face_auth.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import Syncano from 'syncano-server';
+
+export default (ctx) => {
+  const {response, users } = Syncano(ctx);
+
+  const {username, token} = ctx.args;
```
Comment (adamwardecki) [inline src/verify_face_auth.js:6]: Do we need the username here and as a param in the yml (same goes for the remove_face_auth)?

 `users.where('user_key', token)`

**Review by adamwardecki — CHANGES_REQUESTED** [review]
Look good! Check my comments.

File: src/verify_face_auth.js (hunk 8 above)
Comment (eshaibu) [reply src/verify_face_auth.js:6]: True, not needed

Comment (eshaibu) [reply src/verify_face_auth.js:6]: Seems searching by `user_key` does not work even after adding it to socket YAML and  setting the `filter_index` to `true`

**Comment by adamwardecki** [comment]
Ok, looks good. You can merge, create a repo in Syncano org on github and publish the changes there. Once done publish the socket to the Syncano registry.

**Pull request merged**